nnstreamer/tensor_filter/tensor_filter_nntrainer.cc

```diff
@@ -45,110 +46,73 @@ static const gchar *nntrainer_accl_support[] = {NULL};
-
-  nntrainer::TensorDim dim;
-  nntrainer_tensor_info_s info_s;
-
-  if (is_input)
-    dim = model->getInputDimension()[0];
-  else
-    dim = model->getOutputDimension()[0];
-
-  if (tensorInfo->info[0].type != _NNS_FLOAT32)
-    throw std::invalid_argument(
-      "only float32 is supported for input and output");
-
-  info_s.rank = NUM_DIM;
-
-  info_s.dims.push_back(tensorInfo->info[0].dimension[NUM_DIM - 1]);
-
-  // Here, we only compare channel, height, width.
-  // Just use batch from info variable, cause it will be updated if it differs.
-  for (unsigned int i = 1; i < NUM_DIM; ++i) {
-    if (tensorInfo->info[0].dimension[i - 1] != dim.getDim()[NUM_DIM - i])
-      throw std::invalid_argument("Tensor dimension doesn't match");
+static nntrainer::TensorDim to_nntr_tensor_dim(const GstTensorInfo *info) {
+  const tensor_dim &d = info->dimension;
+  return {d[3], d[2], d[1], d[0]};
+}
 
-    info_s.dims.push_back(dim.getDim()[i]);
-  }
+NNTrainerInference::NNTrainerInference(const std::string &model_config_) :
+  model_config(model_config_) {
+  loadModel();
+  model->compile();
+  model->initialize();
+  model->readModel();
+}
 
-  if (is_input)
-    input_tensor_info.push_back(info_s);
+const char *NNTrainerInference::getModelConfig() {
+  return model_config.c_str();
 }
 
 void NNTrainerInference::loadModel() {
 #if (DBG)
   gint64 start_time = g_get_real_time();
 #endif
-  if (model_config == nullptr)
-    throw std::invalid_argument("model config is null!");
-
-  model = new nntrainer::NeuralNetwork();
+  model = std::make_unique<nntrainer::NeuralNetwork>();
   model->loadFromConfig(model_config);
 #if (DBG)
   gint64 stop_time = g_get_real_time();
   g_message("Model is loaded: %" G_GINT64_FORMAT, (stop_time - start_time));
 #endif
 }
 
-int NNTrainerInference::getInputTensorDim(GstTensorsInfo *info) {
-  gst_tensors_info_copy(info, &inputTensorMeta);
-  return 0;
-}
-
-int NNTrainerInference::getOutputTensorDim(GstTensorsInfo *info) {
-  gst_tensors_info_copy(info, &outputTensorMeta);
-  return 0;
-}
-
 int NNTrainerInference::run(const GstTensorMemory *input,
                             GstTensorMemory *output) {
 #if (DBG)
   gint64 start_time = g_get_real_time();
 #endif
   std::shared_ptr<nntrainer::Tensor> out;
 
-  std::vector<std::int64_t> d = input_tensor_info[0].dims;
-  nntrainer::Tensor X =
-    nntrainer::Tensor(nntrainer::TensorDim(d[0], d[1], d[2], d[3]),
-                      static_cast<float *>(input[0].data));
+  auto input_dims = getInputDimension();
+  nntrainer::sharedConstTensors inputs;
+  inputs.reserve(input_dims.size());
+
+  const GstTensorMemory *input_mem = input;
+  unsigned int offset = 0;
+  for (auto &id : input_dims) {
+    // do not allocate new, but instead use tensor::Map
+    inputs.emplace_back(MAKE_SHARED_TENSOR(nntrainer::Tensor::Map(
+      static_cast<float *>(input_mem->data), input_mem->size, id, offset)));
+    input_mem++;
+    offset += id.getDataLen() * sizeof(float);
```

Comment:
this is minor thing.
As this tensor_filter is written for nnstreamer, this filter must trust nnstreamer side info as the framework (nntrainer here) can have bugs. Further, the memory here being addressed is allocated by nnstreamer. So, IMO the offset must use the size from the input_dim stored in the filter/GstTensorMemory than getting the dimensions from the framework (nntrainer) and using it. Current method can lead to issues if the framework updates dimensions internally due to some reason.

Reply:
sounds reasonable, I'll change

Reply:
Done